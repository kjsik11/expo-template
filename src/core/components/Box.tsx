import React from "react";
import {
  FlexAlignType,
  StyleProp,
  View,
  ViewProps,
  ViewStyle,
} from "react-native";

import { Spacing, SpacingInNumber } from "core/types";
import {
  calculateSpacing,
  calculateSpacingInNumber,
} from "core/helpers/spacing";

export interface BoxProps extends ViewProps {
  m?: Spacing;
  mt?: Spacing;
  mb?: Spacing;
  mr?: Spacing;
  ml?: Spacing;
  mv?: Spacing;
  mh?: Spacing;

  p?: Spacing;
  pt?: Spacing;
  pb?: Spacing;
  pr?: Spacing;
  pl?: Spacing;
  pv?: Spacing;
  ph?: Spacing;

  height?: Spacing | string;
  width?: Spacing | string;
  border?: SpacingInNumber;
  borderRadius?: SpacingInNumber;
  opacity?: number;

  justifyContent?:
    | "flex-start"
    | "flex-end"
    | "center"
    | "space-between"
    | "space-around"
    | "space-evenly"
    | undefined;
  alignItems?: FlexAlignType;

  flex?: number;
  direction?: "column" | "row";

  backgroundColor?: string;

  customStyle?: any;
}

function Box(props: BoxProps) {
  const style: StyleProp<ViewStyle> = {
    margin: calculateSpacing(props.m),
    marginTop: calculateSpacing(props.mt),
    marginBottom: calculateSpacing(props.mb),
    marginRight: calculateSpacing(props.mr),
    marginLeft: calculateSpacing(props.ml),
    marginVertical: calculateSpacing(props.mv),
    marginHorizontal: calculateSpacing(props.mh),

    padding: calculateSpacing(props.p),
    paddingTop: calculateSpacing(props.pt),
    paddingBottom: calculateSpacing(props.pb),
    paddingRight: calculateSpacing(props.pr),
    paddingLeft: calculateSpacing(props.pl),
    paddingVertical: calculateSpacing(props.pv),
    paddingHorizontal: calculateSpacing(props.ph),

    height: calculateSpacing(props.height),
    width: calculateSpacing(props.width),
    borderWidth: calculateSpacingInNumber(props.border),
    borderRadius: calculateSpacingInNumber(props.borderRadius),
    opacity: calculateSpacingInNumber(props.opacity),

    justifyContent: props.justifyContent,
    alignItems: props.alignItems,

    flex: props.flex,
    flexDirection: props.direction,

    backgroundColor: props.backgroundColor,

    ...props.customStyle,
  };

  return <View {...props} style={style} />;
}

export default Box;
