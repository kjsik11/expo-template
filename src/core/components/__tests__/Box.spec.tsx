import React from "react";
import { Text } from "react-native";
import { render, screen } from "@testing-library/react-native";
import { ReactTestInstance } from "react-test-renderer";

import Box from "core/components/Box";

const TEST_ID_OF_BOX = "box";
const TEXT = "Box Children";
const CHILDREN = <Text>{TEXT}</Text>;

describe("ðŸ«€ Core | ðŸ§© Components | ðŸ“¦ <Box>", function () {
  let box: ReactTestInstance | null = null;

  beforeEach(function () {
    render(<Box children={CHILDREN} testID={TEST_ID_OF_BOX} />);

    box = screen.getByTestId(TEST_ID_OF_BOX);
  });

  it("should render successfully", () => {
    expect(box).toBeDefined();
  });

  it("should render correctly", () => {
    const boxAsJson = screen.toJSON();
    expect(boxAsJson).toMatchSnapshot();
  });

  it("should render with title", () => {
    expect(screen.getByText(TEXT)).toBeTruthy();
  });
});
