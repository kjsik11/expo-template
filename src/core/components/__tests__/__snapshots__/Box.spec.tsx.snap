// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ğŸ«€ Core | ğŸ§© Components | ğŸ“¦ <Box> should render correctly 1`] = `
<View
  style={
    Object {
      "alignItems": undefined,
      "backgroundColor": undefined,
      "borderRadius": undefined,
      "borderWidth": undefined,
      "flex": undefined,
      "flexDirection": undefined,
      "height": undefined,
      "justifyContent": undefined,
      "margin": undefined,
      "marginBottom": undefined,
      "marginHorizontal": undefined,
      "marginLeft": undefined,
      "marginRight": undefined,
      "marginTop": undefined,
      "marginVertical": undefined,
      "opacity": undefined,
      "padding": undefined,
      "paddingBottom": undefined,
      "paddingHorizontal": undefined,
      "paddingLeft": undefined,
      "paddingRight": undefined,
      "paddingTop": undefined,
      "paddingVertical": undefined,
      "width": undefined,
    }
  }
  testID="box"
>
  <Text>
    Box Children
  </Text>
</View>
`;
